examples/hetero/bipartite_sage.py

```diff
@@ -0,0 +1,262 @@
+
+    return u2imat
+
+
+def get_coocur_mat(train_mat, num_neighbors):
+    # Generate the co-occurrence matrix with normalisation and top-k filtering
+    A = train_mat.T @ train_mat
+    n_items = A.shape[0]
+    res_mat = torch.zeros((n_items, num_neighbors))
+    res_sim_mat = torch.zeros((n_items, num_neighbors))
+
+    items_D = torch.sum(A, axis=0).reshape(-1)
+    users_D = torch.sum(A, axis=1).reshape(-1)
+    beta_uD = (1 / torch.sqrt(users_D + 1)).reshape(-1, 1)
+    beta_iD = (1 / torch.sqrt(items_D + 1)).reshape(1, -1)
+    all_ii_constraint_mat = beta_uD @ beta_iD
+
+    for i in range(n_items):
+        row = all_ii_constraint_mat[i] * A[i]
+        row_sims, row_idxs = torch.topk(row, num_neighbors)
+        res_mat[i] = row_idxs
+        res_sim_mat[i] = row_sims
+
+    return res_mat.long(), res_sim_mat.float()
+
+
+def get_i2i_sim_graph(i2imat, i2isimmat):
+    # Generate the i2i graph from the co-occurrence matrix
+    i2i, i2i_sim = [[], []], []
+    for i in range(len(i2imat)):
+        # Filter nodes with no neighbors in co-occurrence matrix
+        if i2imat[i][0] == i and i2isimmat[i][0] > 1e-5:
+            i2i_sim.append(i2isimmat[i][1:])
+            for j in i2imat[i]:
+                i2i[0].append(i)
+                i2i[1].append(j)
+
+    return torch.tensor(i2i), torch.stack(i2i_sim)
+
+
+u2i_mat = to_u2i_mat(train_data.edge_index_dict[('user', 'rates', 'movie')],
+                     train_data['movie'].x.shape[0],
+                     train_data['user'].x.shape[0])
+i2i_mat, i2i_sim_mat = get_coocur_mat(u2i_mat, 9)
+i2i, _ = get_i2i_sim_graph(i2i_mat, i2i_sim_mat)
+
+# Add the generated i2i graph for high-order information
+train_data[('movie', 'sims', 'movie')].edge_index = i2i.to(device)
+val_data[('movie', 'sims', 'movie')].edge_index = i2i.to(device)
+test_data[('movie', 'sims', 'movie')].edge_index = i2i.to(device)
+
+# We have an unbalanced dataset with many labels for rating 3 and 4, and very
+# few for 0 and 1. Therefore we use a weighted MSE loss.
+if args.use_weighted_loss:
+    weight = torch.bincount(train_data['user', 'movie'].edge_label)
+    weight = weight.max() / weight
+else:
+    weight = None
+
+
+def weighted_mse_loss(pred, target, weight=None):
+    weight = 1. if weight is None else weight[target].to(pred.dtype)
+    return (weight * (pred - target.to(pred.dtype)).pow(2)).mean()
+
+
+class ItemGNNEncoder(torch.nn.Module):
+    def __init__(self, hidden_channels, out_channels):
+        super().__init__()
+
+        self.conv1 = SAGEConv(-1, hidden_channels)
+        self.conv2 = SAGEConv(-1, hidden_channels)
+        self.lin1 = Linear(hidden_channels, hidden_channels)
+        self.lin2 = Linear(hidden_channels, out_channels)
+
+    def reset_parameters(self):
+        self.conv1.reset_parameters()
+        self.conv2.reset_parameters()
+        self.lin1.reset_parameters()
+        self.lin2.reset_parameters()
+
+    def forward(self, x_dict, edge_index_dict):
```

Comment:
```suggestion
    def forward(self, x, edge_index):
```
No need to work with dicts here.

Reply:
fixed.